### PR TITLE
allow multiple NADs of the same network on a Pod

### PR DESCRIPTION
**Pull request by cathy-zhou**
Signed-off-by: Yun Zhou <yunz@nvidia.com>

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/util/multi_network.go
Comment (trozet) [inline go-controller/pkg/util/multi_network.go:311]: wouldn't a pod always want the primary network, or not have any choice whether or not they want the primary network?

Comment (cathy-zhou) [reply go-controller/pkg/util/multi_network.go:311]: I thought we discussed that, and we conclude that we assume the default(primary network) is OVN, regardless if its "v1.multus-cni.io/default-network" annotation actually request a non-OVN network (that is not supported).

File: go-controller/pkg/util/multi_network.go
Comment (cathy-zhou) [reply go-controller/pkg/util/multi_network.go:302]: this was suggested once, but PodWantsNetwork() was already used to check this pod is not hostNetworked.

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (jcaamano) [inline go-controller/pkg/ovn/base_network_controller_secondary.go:15]: nit: whats wrong with just not having an alias here and just using `errors`?

File: go-controller/pkg/ovn/base_network_controller_secondary.go
Comment (cathy-zhou) [reply go-controller/pkg/ovn/base_network_controller_secondary.go:202]: Hmm, true, but isn't that a problem even before the multi-homing support. I will think about a fix.

**Comment by cathy-zhou** [comment]
@trozet @jcaamano I've made the changes based on your comments. Please take a look. Thanks!

File: go-controller/pkg/util/multi_network.go
Comment (jcaamano) [inline go-controller/pkg/util/multi_network.go:432]: Just had this thought. I guess that we are assuming that a host network pod cannot be attached to secondary networks. But API wise there is nothing preventing the user from attempting such a thing. Are we checking this somewhere and returning an error?

File: go-controller/pkg/ovn/pods.go
Comment (jcaamano) [inline go-controller/pkg/ovn/pods.go:153]: nit: would it be better to rather check this is empty or has a single item and return an error otherwise.

Additionally, are we validating somewhere that the default-network annotation actually points to a NAD attaching to the primary network. I guess this should fail somewhere with a meaningful error.

**Review by jcaamano — COMMENTED** [review]
Just a couple of nits more.
I will leave this to be merged by @trozet on Friday, I guess that's not a significant delay and he can check out his comments as well.
BTW @cathy-zhou @maiqueb we can probably use an e2e test for this case. I can help next week if you don't have the time.

File: go-controller/pkg/util/multi_network.go
Comment (cathy-zhou) [reply go-controller/pkg/util/multi_network.go:432]: I don't quite understand this comment. If a pod is host-networked, I think even multus-cni won't be triggered, as the result, multus won't be able to start cni for secondary network either.

File: go-controller/pkg/ovn/pods.go
Comment (cathy-zhou) [reply go-controller/pkg/ovn/pods.go:153]: then it is probably best for nad controller to monitor default nad as well, and let the controller also has nadNames sync.Map so it can be validated by HasNAD() function. That is how it is done in downstream as well.

If you think that is the way to go, I'd rather leave it as a later PR. Before we don't valid NAD for default network either, this PR does not make it worse.

Comment (cathy-zhou) [reply go-controller/pkg/ovn/pods.go:153]: for this first part fo the comments, the way how default NAD is defined determined that the networkMap is either empty or has single NAD (either "v1.multus-cni.io/default-network" is defined or not). Do we still need to check it here?

**Review by trozet — CHANGES_REQUESTED** [review]
Thanks for changing the util.PodWantsHostNetwork. I think this is much clearer. However, I still think PodWantsNetwork should be renamed to something like GetPodNADToNetworkMapping.

**Comment by cathy-zhou** [comment]
> Thanks for changing the util.PodWantsHostNetwork. I think this is much clearer. However, I still think PodWantsNetwork should be renamed to something like GetPodNADToNetworkMapping.

@trozet updated. please take a look. Thanks!

File: go-controller/pkg/util/multi_network.go
Comment (jcaamano) [reply go-controller/pkg/util/multi_network.go:432]: Yeah. So if a user attempts to run a pod with `hostNetwork: true` and annotated with `k8s.v1.cni.cncf.io/networks` (or `v1.multus-cni.io/default-network`) pointing to a NAD with an ovn-k secondary network, we simply ignore the annotation instead of reporting an error somewhere for that pod given that we cannot comply with what the user is requesting.

File: go-controller/pkg/ovn/pods.go
Comment (jcaamano) [reply go-controller/pkg/ovn/pods.go:153]: > then it is probably best for nad controller to monitor default nad as well, so the default network controller also has nadNames sync.Map so it can be validated by HasNAD() function. That is how it is done in downstream as well.
> 
> If you think that is the way to go, I'd rather leave it as a later PR. Before we don't valid NAD for default network either, this PR does not make it worse.

I think it could be simpler than that, we could just check that the NAD is of primary type (whatever criteria we are using to determine that) in `GetK8sPodDefaultNetworkSelection`. We can leave it for a later PR if you think it does not fit this PR.

Comment (jcaamano) [reply go-controller/pkg/ovn/pods.go:153]: > for this first part fo your comment, the way how default NAD is defined determined that the networkMap is either empty or has single NAD (either "v1.multus-cni.io/default-network" is defined or not). Do we still need to check it here?

I just meant adding something like
```
if len(networkMap) > 1 {
  return fmt.Errorf(....)
}
```
but I guess we can assume we will only ever get one network for the default as there is no possible way API wise to specify multiple different NADs for the default network.

Comment (trozet) [reply go-controller/pkg/ovn/pods.go:153]: i think for the sake of this code we should have an error check like the above. We shouldn't depend on API or code from other places to catch an error in this function. It's better to have more robust error checking where possible.

**Comment by trozet** [comment]
hey @cathy-zhou looks good, just would like you to add a check for more than one network as @jcaamano mentioned

**Comment by cathy-zhou** [comment]
> hey @cathy-zhou looks good, just would like you to add a check for more than one network as @jcaamano mentioned

updated. thanks!

**Comment by jcaamano** [comment]
So these two nits remaining:
* https://github.com/ovn-org/ovn-kubernetes/pull/3360#discussion_r1097251929
* https://github.com/ovn-org/ovn-kubernetes/pull/3360#discussion_r1097251929

If we rather keep things as they are or we we want to to address it on a separate PR, LGTM.

**Comment by cathy-zhou** [comment]
> So these two nits remaining:
> 
> https://github.com/ovn-org/ovn-kubernetes/pull/3360#discussion_r1097251929
> https://github.com/ovn-org/ovn-kubernetes/pull/3360#discussion_r1097259629
> If we are good with it or want to defer to a later PR, LGTM.
for the second one, while I'd rather to delay it to a later PR,  I am also not sure what to do if the network selection elements for the default network is of second network, shall we return failure. But If it is of non-OVN network, ovnkube-master also just blindly assumes it is managd by OVN , right? 

For the first one, the user has conflict requests: it requested host-network Pod, while also requesting some network attachment. CNI that it requested is not even called, I don't think CNI is responsible to report the error.

**Review by trozet — APPROVED** [review]

**Comment by maiqueb** [comment]
@cathy-zhou I'm trying to add an e2e test to assert this works, but I've realized I don't know how to use it.

Can you please update this PR's description with a reference yaml on how to configure such a scenario ?...

**Pull request merged**
